pynitrokey/start/upgrade_by_passwd.py

```diff
@@ -427,116 +465,116 @@ def start_update(regnual, gnuk, default_password, password, wait_e, keyno, verbo
             update_done = True
             break
+
+        # @todo: add proper exceptions (for each case) here
         except ValueError as e:
-            logger.exception('Error while running update')
-            str_factory_reset = 'Please "factory-reset" your device to ' \
-                                'continue (this will delete all user data from the device) ' \
-                                'and try again with PIN="12345678".'
-            if 'No ICC present' in str(e):
+            local_print("error while running update")
+            str_factory_reset = "Please 'factory-reset' your device to " \
+                "continue (this will delete all user data from the device) " \
+                "and try again with PIN='12345678'"
+
+            if "No ICC present" in str(e):
                 kill_smartcard_services()
-                # local_print('*** Please run update tool again.')
+                local_print("retrying...")
+
             else:
-                local_print('*** Could not proceed with the update.')
-                local_print('*** Found error: {}'.format(str(e)))
-                # FIXME run factory reset here since data are lost anyway
+                # @fixme run factory reset here since data are lost anyway (rly?)
                 if str(e) == ERR_EMPTY_COUNTER:
-                    local_print('*** Device returns "Attempt counter empty" error for Admin PIN.'
-                          + ' ' + str_factory_reset
-                          )
+                    local_critical("- device returns: 'Attempt counter empty' "
+                                   "- error for Admin PIN", str_factory_reset, e)
+
                 if str(e) == ERR_INVALID_PIN:
-                    local_print('*** Device returns "Invalid PIN" error.'
-                          + ' ' + str_factory_reset)
-                break
+                    local_critical("- device returns: 'Invalid PIN' error",
+                                   "- please retry with correct PIN", e)
         except Exception as e:
-            # unknown error, bail
-            local_print('*** Found unexpected error: {}'.format(str(e)))
-            break
+            local_critical("unexpected error", e)
 
     if not update_done:
-        local_print()
-        local_print('*** Could not proceed with the update. Please execute one or all of the following and try again:\n'
-              '- reinsert device to the USB slot;\n'
-              '- run factory-reset on the device;\n'
-              '- close other applications, that possibly could use it (e.g. scdaemon, pcscd).\n')
-        exit(1)
+        local_critical("",
+            "Could not proceed with the update",
+            "Please execute one or all of the following and try again:",
+            "- re-insert device to the USB slot",
+            "- run factory-reset on the device",
+            "- close other applications, which could use it (e.g., scdaemon, pcscd)")
```

Comment:
Device should never be removed during the update procedure. This should be executed only if the update has not started at all. To detect and correct the help strings.